### ISSUE TITLE
Show the commit header in the list of branches to delete

### ISSUE THREAD

**Issue by jack1142** [hartwork/git-delete-merged-branches#23]
It would be very useful, if the "You are about delete 1 local/remote branch(es):" list would show the commit header, similarly to the output of `git branch -v`. Sometimes the branch names don't immediately tell you what that branch contains and showing commit header could quickly help with that issue.

**Comment by hartwork** [comment]
Hi @jack1142 ,

this would certainly not be too hard to implement but I have concerns that it is the right move: Human attention is limited and if something new eats attention, the rest gets less: Cognitive load is increased for projects with good self-explaining branches without any gain in return. Also, I found `git branch -v` less helpful than branch names themselves in the cases I have seen, because the latest commit can be something about a detail and does not necessarily summarize the big picture, especially if you care about small atomic commits that e.g. separate bugfixes from refactorings. Plus bad branch names is a cultural problem that's either easy to fix in a healthy environment or a good indicator that the environment struggles with even easy-to-fix issues (..). Plus it's easy to hit Ctrl+C, inspect one or two branches with a better chance at getting to the right information and then run the tool again.

So to summarize my impression right now is that it's not a *big* issue, has good easy workarounds, can be fixed outside of the tool, is hard to address really well with little space for everyone, and affects the general case in a negative way.

What do you think?

**Comment by jack1142** [comment]
For me, even great branch names lose the information after you haven't done any work with them for a while.

I get the concern of increased cognitive load, but I think that this in reality is only true when you use the tool for the first time as you need to learn what each part of the interface does. Later, your mind will just "automatically" ignore the part of the interfaces it has no use for and only look at them when it needs them.
But with that said, you're probably right that this is unneeded in most of the times. I don't think this should mean that we shouldn't have a way to see it. There could for example be an option to show more verbose info when it asks you whether you want to delete the branches (right now it's just `y/n` prompt).

I know that you said that "it's easy to hit Ctrl+C, inspect one or two branches", but imo this requires more work and depending on how user chooses to do it, it can also require context-switching, which will distract the user even more. This is the exact reason why I'm proposing it - I'm losing more time than I would want on keeping my repository clean. `git-delete-merged-branches` is already a big step in making that easier for me, and the time I would gain with this feature wouldn't be as big as the gain I had when switching from manual cleaning to this tool, but it's still a gain.

**Comment by hartwork** [comment]
Thanks for providing more information!  I've been thinking about this more.  `git-delete-merged-branches` works with the assumption that the general case is that you want to remove all merged branches (with regard to all selected target branches) and that you blacklist any branches that you know you want to keep.  If that doesn't cover your case, I would be curious to hear why.  So the display of branches to delete is meant to be a safety gate, nothing more.  The fix to longterm branches with names that do not explain themselves two weeks later is twofold in my eyes: keep branches short-lived and ensure good enough naming; because long-lived branches are trouble anyway and better names should be easy to do.  Once you have good enough names, there should be no need for `git branch -v` info any more.  Am I missing something?

**Comment by jack1142** [comment]
> So the display of branches to delete is meant to be a safety gate, nothing more.
Yes, I treat it exactly as that. I prefer to put limited trust in the tooling, so I want to be sure that it doesn't do something I don't want it to and if I don't know what the branch it tries to delete is, I need to check it.

> keep branches short-lived and ensure good enough naming; because long-lived branches are trouble anyway and better names should be easy to do.
IMO, this (keeping branches short-lived) may be easy/easier in a work environment, where you get paid for your time, but it can be a lot more troublesome in open source projects, where you can't really know when someone will have time to review your work.

But even with that said after thinking a bit more about it, I don't know whether `git branch -v` is really that good solution here. I know that it generally helps me with figuring out what the branch does, but you are right that it can highly depend on your last commit message. And I don't know how else could you get more information right from the tool without hurting the general case. I'm guessing that what I mentioned with showing more verbose info could be expanded even further if you could select specific branch you want to see info for, but that begins to complicate this feature request more.

**Comment by hartwork** [comment]
Thanks for elaborating further!

> > So the display of branches to delete is meant to be a safety gate, nothing more.
> 
> Yes, I treat it exactly as that. I prefer to put limited trust in the tooling, so I want to be sure that it doesn't do something I don't want it to and if I don't know what the branch it tries to delete is, I need to check it.

Did you see the [notes regarding safety in the readme](https://github.com/hartwork/git-delete-merged-branches#safety)?  Does anything come to mind that could help with trust in this tool for you or others?


> > keep branches short-lived and ensure good enough naming; because long-lived branches are trouble anyway and better names should be easy to do.
> 
> IMO, this (keeping branches short-lived) may be easy/easier in a work environment, where you get paid for your time, but it can be a lot more troublesome in open source projects, where you can't really know when someone will have time to review your work.

So branch names are not easy to improve on in a pull-request flow a variety of contributors, I understand, good point.

> But even with that said after thinking a bit more about it, I don't know whether `git branch -v` is really that good solution here. I know that it generally helps me with figuring out what the branch does, but you are right that it can highly depend on your last commit message. And I don't know how else could you get more information right from the tool without hurting the general case. I'm guessing that what I mentioned with showing more verbose info could be expanded even further if you could select specific branch you want to see info for, but that begins to complicate this feature request more.

Then I guess the question if there is some middle ground with providing additional but limited details per branch that are on-point enough to allow a decision on that branch when the branch name alone was not enough yet the extra details provided are enough to resolve need for manual in-depth research on that branch, at least most of the time.  What do you think?

**Comment by jack1142** [comment]
> Did you see the notes regarding safety in the readme? Does anything come to mind that could help with trust in this tool for you or others?

Yes, I read them and they are partially why I'm using this tool - thanks to how it works, I am sure it won't delete a branch I don't want it to. But that is not the issue here, the issue is that I don't know if I want to delete *that specific branch on the list* because I don't have enough information.

> Then I guess the question if there is some middle ground with providing additional but limited details per branch that are on-point enough to allow a decision on that branch when the branch name alone was not enough yet the extra details provided are enough to resolve need for manual in-depth research on that branch, at least most of the time. What do you think?

Yep, this is what I'm looking for, I proposed solution that shows data like `git branch -v` because I couldn't think of a better alternative (and it generally works **for me**, so I didn't think much about it).

**Comment by hartwork** [comment]
> [..] the issue is that I don't know if I want to delete _that specific branch on the list_ because I don't have enough information.

Can you give an example for a reason to keep a merged branch?

The only case that I'm aware of so far is multiple release-like branches e.g. `stable`, `release`, `dev` or so. That's what the exclude feature is for. The set of these branches supposedly rather small and rather static. Is your scenario different?

**Comment by jack1142** [comment]
> Can you give an example for a reason to keep a merged branch around?

That's what I meant when I said that I put limited trust in tools I use. I don't (yet) trust that squash merge detection doesn't have any issues (I haven't seen any, but this is a rather new tool) and so I always want to confirm that it indeed only removes what it should.

Since this is just another safety measure, maybe it should just be available as a flag then?

**Comment by hartwork** [comment]
> That's what I meant when I said that I put limited trust in tools I use. I don't (yet) trust that squash merge detection doesn't have any issues (I haven't seen any, but this is a rather new tool) and so I always want to confirm that it indeed only removes what it should.

Above^^ you're saying it's the tool and earlier you said it's the branches:

> [T]hanks to how it works, I am sure it won't delete a branch I don't want it to. But that is not the issue here, the issue is that I don't know if I want to delete _that specific branch on the list_ because I don't have enough information.

So which is it? Please clarify.


> Since this is just another safety measure, maybe it should just be available as a flag then?

Detection of squash merges is already guarded behind `--effort=3` so it's only active when you ask for it:

```console
# git-delete-merged-branches --help | fgrep -A4 ' --effort'
  --effort LEVEL        level of effort to put into finding merged branches;
                        level 1 uses nothing but "git branch --merged", level
                        2 adds use of "git cherry", level 3 adds use of "git
                        cherry" on temporary squashed copies (default level:
                        2)
```

**Comment by jack1142** [comment]
> Above^^ you're saying it's the tool and earlier you said it's the branches:

I'm sorry, I meant the same thing with both here - I want to be sure that *the tool* doesn't remove *some branch* it shouldn't because of incorrect detection of squash merge (or possibly, although very unlikely, me having *some* reason to not remove it; I don't have anything specific in mind here though).

> Detection of squash merges is already guarded behind --effort=3 so it's only active when you ask for it:

I do want it to detect squash merges. I don't want it to delete something that wasn't actually a squash (or other kind of) merge if it were to detect it incorrectly (false-positive). I know that the tool is supposed to do it properly, but I want to lean on the safe side here.

**Comment by hartwork** [comment]
Okay. Can we agree that adding `git branch -v` is not all positive right now, that we'll close the ticket for now, and that you're free to open a new bug report if you feel need to again in the future? The ticket does not seem actionable to me right now and I would rather not have it sit here open for weeks with an unclear future. What do think?
